### ISSUE TITLE
[maybe bug]Incorrect Handling of Input Sequences and Labels in DataCollatorForCausalLM

### ISSUE THREAD

**Issue by oplex151** [jzhang38/TinyLlama#205]
https://github.com/jzhang38/TinyLlama/blob/bf122247c486b6b897050e98cbb7bedae8eeba73/sft/finetune.py#L287

## Description:
There is a potential issue identified in the DataCollatorForCausalLM class. Specifically, the logic for generating input_ids and labels does not correctly handle the offset (or shift) required for causal language modeling (CLM). In CLM, each token in the input sequence should predict the next token in the sequence. However, the current implementation may lead to incorrect alignment between input_ids and labels, which can hinder the model's ability to learn effectively.

## Code Snippet:

```
for tokenized_source, tokenized_target in zip(
    tokenized_sources_with_prompt['input_ids'],
    tokenized_targets['input_ids']
):
    if not self.predict_with_generate:
        input_ids.append(torch.tensor(tokenized_source + tokenized_target))
        if not self.train_on_source:  # Potential bug here; x and y should have an offset
            labels.append(
                torch.tensor([IGNORE_INDEX for _ in range(len(tokenized_source))] + copy.deepcopy(tokenized_target))
            )
        else:
            labels.append(torch.tensor(copy.deepcopy(tokenized_source + tokenized_target)))
```
## Expected Behavior:
For causal language models, the correct approach is to ensure that each token in `input_ids` corresponds to the next token in `labels`. This means that for a given sequence, the label for each position should be the token that follows it in the sequence. The last token in `input_ids` should correspond to an IGNORE_INDEX in `labels`, as there is no subsequent token to predict.

## Suggested Fix:
Adjust the logic for generating labels to account for the necessary offset between `input_ids` and `labels.` Here is the corrected code snippet:
```
for tokenized_source, tokenized_target in zip(tokenized_sources_with_prompt['input_ids'], tokenized_targets['input_ids']):
            if not self.predict_with_generate:
                combined_input = tokenized_source + tokenized_target
                input_ids.append(torch.tensor(combined_input))
                
                if not self.train_on_source:
                    label = [self.IGNORE_INDEX] * len(tokenized_source) + deepcopy(tokenized_target)
                else:
                    label = deepcopy(combined_input)

                # Adjust labels to account for the shift between input and output
                labels.append(torch.tensor(label[1:] + [self.IGNORE_INDEX]))  # Last token has no corresponding label
                
            else:
                input_ids.append(torch.tensor(tokenized_source))

```
